### PR TITLE
Strong invariants in polymorphic definitions

### PR DESCRIPTION
**Pull request by ppedrot**
This PR enforces in the kernel the following invariant for polymorphic definitions: both internal constraints (as generated by forcing the body) and side-effect constraints (as generated by side-effect inlining) must be empty. This allows to ensure that there is no messy universe context union in the kernel, and in particular that the universe bounds provided by the user are preserved.

As such, it is an alternative of #735. I'm suspecting the latter was not thoroughly tested, or at least at the time a few tests were missing, because this PR required a few subtle changes in the implementation of abstract.

This PR breaks backward compatibility in a few very uncommon corner cases. For instance, HoTT breaks but can be fixed with a 6-line patch fiddling with universe binders.

This PR depends on #6469.

### CODE REVIEW

**Review by SkySkimmer — COMMENTED** [review]
I don't understand the side effect stuff very well so not hitting the official approve button, but I'm always happy to have less useless universes.

**Review by mattam82 — APPROVED** [review]
I really like it, the model is simpler and not too restrictive. The code LGTM too.

**Comment by ppedrot** [comment]
For some reason I don't understand, Travis has not picked my HoTT branch but the official one and has thus miserably failed. But it does work on my machine, no need to perform a full compilation of our benches again...

**Comment by SkySkimmer** [comment]
>for overlay in '${ci_dir}/user-overlays/*.sh'

The overlay file must end in `.sh`.

**Comment by ppedrot** [comment]
Gaaah! I'll force-push a fix.

**Comment by maximedenes** [comment]
> For instance, HoTT breaks but can be fixed with a 6-line patch fiddling with universe binders.

@ppedrot Is your HoTT patch backward compatible? Can you open a PR on HoTT?

**Comment by ppedrot** [comment]
@maximedenes The HoTT patch is not backward compatible, even though I think I can make it so by using the kind of tricks @SkySkimmer mentioned. That said, I don't know whether HoTT actually insists on backward compatibility that much (they only support 8.7) and furthermore the features that would make it more robust to changes, namely explicit universe constraints, were only introduced in 8.8.

**Comment by maximedenes** [comment]
If that's ok with the HoTT devs, then fine! Can you open a PR there? Unless I missed it.

**Comment by SkySkimmer** [comment]
>That said, I don't know whether HoTT actually insists on backward compatibility that much (they only support 8.7)

There's https://github.com/HoTT/HoTT/pull/906#pullrequestreview-78921048 (@JasonGross)
>The compatibility hack is ugly, but I think it's worth it to support 8.7 until 8.8 comes out if we can.

**Comment by maximedenes** [comment]
@ppedrot It seems the HoTT devs are asking for a backward compatible fix instead.

**Comment by ppedrot** [comment]
Ready to merge.

**Comment by Zimmi48** [comment]
There is another reason to prefer backward compatible fixes: it is that in general for libraries, we test the same branch for 8.7 and 8.8. Of course, if upstream authors were willing to drop support for 8.7, we would just have to fix this by pinning what we test in 8.7 to a specific tag, like I did recently on the 8.6 branch.

File: kernel/term_typing.ml [hunk 1]
```diff
@@ -300,21 +300,28 @@ let infer_declaration (type a) ~(trust : a trust) env (dcl : a constant_entry) =
...
+        (** Ensure not to generate internal constraints in polymorphic mode.
+            The only way for this to happen would be that either the body
+            contained deferred universes, or that it contains monomorphic
+            side-effects. The first property is ruled out by upper layers,
+            and the second one is ensured by the fact we currently
+            unconditionally export side-effects from polymorphic definitions,
+            i.e. [trust] is always [Pure]. *)
+        let () = assert (Univ.ContextSet.is_empty ctx) in
+        let env = push_context ~strict:false uctx env in
+        let sbst, auctx = Univ.abstract_universes uctx in
```
Comment (maximedenes) [inline kernel/term_typing.ml]: Should this be `subst`?

**Comment by maximedenes** [comment]
It seems this PR interacts with other universe-related PRs, so it will need a rebase (IIUC).

File: kernel/term_typing.ml (hunk 1 above)
Comment (ppedrot) [reply kernel/term_typing.ml]: You mean just α-renaming?

**Comment by maximedenes** [comment]
@ppedrot Can you rebase? I think I observed the compilation was failing, because of interaction with other PRs, even if there is no conflict. Thanks!

**Comment by ppedrot** [comment]
Rebased, should work now.
